### PR TITLE
Add `backend` concept to `torch.export` API

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Related to https://github.com/pytorch/pytorch/issues/109131

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109650
* __->__ #109649

This PR proposes a way to allow ``torch.export.export`` API to support
multiple backends, similar in principle to the ``torch.compile`` API.

With the new API, one backend can produce, for example, ``GraphModule`` (default backend?)
while the ``onnx`` backend can produce a ``onnx.ModelProto``, unifying all export APIs
from different vendors into one cohesive API.

Backend implementation are separate as it is today, and users could register
custom backends through ``register_backend`` or
list the available ones through ``list_backends``

Below is an example on how a model would be exported with the new API

```python
import torch

def f(x, y):
    return x[0] + y

inp = ([torch.ones(1, 3)], torch.ones(1, 3))

exported_program = torch.export.export(f, inp, backend="dynamo", options={})  # ``backend`` and ``options`` are optionals
output = exported_program(*inp)
```

Besides the main `torch.export.export` API, there are others that need to be discussed:

* `torch.export.load` and `torch.export.save` are also backend specific and ONNX backend, for example, could also implement the same functionalities
* `constrain_as_size`, `constrain_as_value`, `dynamic_dim` that may or may not be reused by other backends

cc @ezyang @gchanan @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
Hi @thiagocrepaldi 

I would like to keep the torch.export() API simple and separated from other "ConvertToOtherIR" functionality. 
A preferable way to get to onnx is 
```
export_program  = torch.export(... )
onnx_program = torch.onnx.convert(export_program)
```

There are multiple IR converters interested in consuming torch.export-ed graph, including onnx, stablehlo. 
Having each IR maintaining it's own converter and API would save us a lot of trouble. 
- You don't want your change to accidentally break other converter path. 
- You can design your own API, introduce configs specific to your IR

**Comment by thiagocrepaldi** [comment]
> Hi @thiagocrepaldi
> 
> I would like to keep the torch.export() API simple and separated from other "ConvertToOtherIR" functionality. A preferable way to get to onnx is
> 
> ```
> export_program  = torch.export(... )
> onnx_program = torch.onnx.convert(export_program)
> ```
Hi @SherlockNoMad, Thanks for the feedback! we implemented something similar through `torch.onnx.dynamo_export` for PT 2.1 release. We didn't use `ExportedProgram` as input for the exporter, but instead we leveraged `torch._dynamo.export` internally. To make a switch, we need to understand how `ExportedProgram` affects vendors leveraging Dynamo under the hood to export their model. There are some assumptions, such as lifting parameters and buffers to inputs, that might affect the final IR in undesired ways.

All in all, the `ExportedProgram` as input for the onnx exporter and any other export API is certainly worth considering, especially because `torch.export` is a public API as of #106904. However, we feel this is an orthogonal discussion to the core of the proposal because vendors can use the proposed API with or without using `ExportedProgram` as input. 

The main goal of this discussion is more about user experience and equal API visibility than a technical challenge 😊


> There are multiple IR converters interested in consuming torch.export-ed graph, including onnx, stablehlo. Having each IR maintaining it's own converter and API would save us a lot of trouble.

By having different vendors implementing different APIs leveraging dynamo under the hood, PyTorch users would have to learn a bunch of different APIs - each with their own philosophy, to experiment and find the best solution for their needs. This might serve as an adoption barrier for users. 

Besides, the API segmentation will only increase as more vendors start leveraging all the goodness of dynamo. 

Our proposal is to create a single entry-point for all vendors that want to leverage Dynamo to export their PyTorch model for whatever purpose. Once the user learns about `torch.export(model)`, it is trivial for users to try `torch.export(model, backend="onnx", options={"onnx_option1": True})` or `torch.export(model, backend="stablehlo", options={"stablehlo_advanced": True})`, etc. This also helps vendors to drive their backend design through a standardized interface, not mentioning the ease of automation due to the uniform syntax. 

A standard API for all creates a better UX as users need only to know one Interface to experiment with many flavors and create a plain field for vendors to expose their engine. 

> * You don't want your change to accidentally break other converter path.

The proposed API does not allow one backend to break another. Each backend is a separate implementation which doesn't interfere with one another, and no changes are needed on the `torch.export` to extend it with extra backends. #109649 implements this proposal as PoC and updates `export` to be a simple dispatch engine to a domain-specific backend: 

```python 
def export( 
    f: Callable, 
    f_args: Tuple[Any, ...], 
    f_kwargs: Optional[Dict[str, Any]] = None, 
    backend: Union[str, Callable] = "dynamo", 
    options: Optional[Dict[str, Any]] = None, 
) -> ExportedProgram: 

    if not options: 
        options = {} 
    export_fn = torch.export.backends.registry._lookup_backend(backend) 
    return export_fn(f, f_args, f_kwargs, **options) 
``` 

As part of the API, `register_backend` and `list_backends` are provided to allow users to add their own thing, keeping the `torch.export` implementation the same, mitigating risks. 

As an example, vendors would implement a backend and a ExportedProgram following this template: 

```python 
@register_backend(name="backend_name") 
def my_backend_implementation( 
    model: Union[torch.nn.Module, Callable], 
    model_args, 
    model_kwargs=None, 
    options: Optional[Any] = None,  # Can be a Dict[str, Any] or any other type 
) -> MyBackendExportedProgram: 

class MyBackendExportedProgram(ExportedProgram): 
    def __call__(self, *args: Any, options=None, **kwargs: Any) -> Any: 
        pass 
``` 

For ONNX, for example, the backend is registered as follows (part of #109650) 
 
```python 
@register_backend(name="onnx") 
def _torch_export_onnx_backend( 
    model: Union[torch.nn.Module, Callable], 
    model_args, 
    model_kwargs=None, 
    export_options: Optional[ExportOptions] = None, 
) -> ONNXExportedProgram: 

    model_kwargs = model_kwargs or {} 
    return torch.onnx.dynamo_export( 
        model, *model_args, export_options=export_options, **model_kwargs 
    ) 
``` 

The `torch.export` was extended but its code is unchanged. This is basically the same concept used by `torch.compile`, ported back to `torch.export` to provide the same flexibility

> * You can design your own API, introduce configs specific to your IR

Indeed, each backend does have their own `options`; the proposed API covers that requirement, as shown through `my_backend_implementation` example above

**Comment by SherlockNoMad** [comment]
`torch._dynamo.export` was not intended to be used as public API, and there is no stability guarantee on this, and we wouldn't encourage vendors/converters to directly integrate with dynamo.export(). 

torch.export() aims to produce a “functional graph with stable IR”, which we believe is a desirable property for many vendors and backends. 

In addition, torch.export() does a lot more than dynamo.export(). To name a few:
- functionalization.
- lowering from torch ops to aten ops (aten ops are a lot more normalized than torch ops) 
- decompose aten ops to a smaller opset (core aten ops) 
You will be missing the goodness of these steps if you directly convert from dynamo.export().

**Comment by thiagocrepaldi** [comment]
> `torch._dynamo.export` was not intended to be used as public API, and there is no stability guarantee on this, and we wouldn't encourage vendors/converters to directly integrate with dynamo.export().
> 
> torch.export() aims to produce a “functional graph with stable IR”, which we believe is a desirable property for many vendors and backends.
> 
> In addition, torch.export() does a lot more than dynamo.export(). To name a few:
> 
> * functionalization.
> * lowering from torch ops to aten ops (aten ops are a lot more normalized than torch ops)
> * decompose aten ops to a smaller opset (core aten ops)
>   You will be missing the goodness of these steps if you directly convert from dynamo.export().

Thank you. We will track the exporter implementation through https://github.com/pytorch/pytorch/issues/109889, which is a related but separate topic.

**Comment by justinchuby** [comment]
> * decompose aten ops to a smaller opset (core aten ops)
>   You will be missing the goodness of these steps if you directly convert from dynamo.export().

torch.onnx needs a fine control on which passes to run. For example, we would like to capture ops beyond core aten, which we have implementations for, by providing a decomposition table.

**Comment by thiagocrepaldi** [comment]
> > * decompose aten ops to a smaller opset (core aten ops)
> >   You will be missing the goodness of these steps if you directly convert from dynamo.export().
> 
> torch.onnx needs a fine control on which passes to run. For example, we would like to capture ops beyond core aten, which we have implementations for, by providing a decomposition table.

Let's track this at https://github.com/pytorch/pytorch/issues/109889

This is not what this PR is about :) For this proposal, we can assume each backend is a black box (which does the "right thing")

**Comment by gmagogsfm** [comment]
> torch.onnx needs a fine control on which passes to run. For example, we would like to capture ops beyond core aten, which we have implementations for, by providing a decomposition table.

This is in the works, backend would be able to control decomposition table more without adding the `backend` argument.

Current PoR:

torch.export() generates full ATen op set without any decomposition, then backend API takes ExportedProgram and use export utilities to run a decomposition table they prefer internally.
